debezium-core/src/main/java/io/debezium/transforms/ExtractNewRecordState.java

```diff
@@ -143,6 +164,61 @@ public R apply(final R record) {
         }
     }
 
+    private R addSourceFields(String[] addSourceFields, R originalRecord, R unwrappedRecord) {
+        // Return if no source fields to add
+        if(addSourceFields == null) {
+            return unwrappedRecord;
+        }
+        
+        final Struct value = requireStruct(unwrappedRecord.value(), PURPOSE);
+        Struct source = ((Struct) originalRecord.value()).getStruct("source");
+        
+        // Get the updated schema from the cache, or create and cache if it doesn't exist. Use the schema of the original unflattened
+        // record which controls both the unwrapped record's schema, and contains the source schema
+        Schema updatedSchema = schemaUpdateCache.get(((Struct) originalRecord.value()).schema());
+        if (updatedSchema == null) {
+            updatedSchema = makeUpdatedSchema(value.schema(), source.schema(),  addSourceFields);
+            schemaUpdateCache.put(value.schema(), updatedSchema);
+        }   
+
+        // Create the updated struct
+        final Struct updatedValue = new Struct(updatedSchema);
+        for (org.apache.kafka.connect.data.Field field : value.schema().fields()) {
+            updatedValue.put(field.name(), value.get(field));
+        }
+        for(String sourceField : addSourceFields) {
+            if (source.schema().field(sourceField) != null) {
+                updatedValue.put(ExtractNewRecordStateConfigDefinition.METADATA_FIELD_PREFIX + sourceField, source.get(sourceField));
+            }
+        }
+
+        return unwrappedRecord.newRecord(
+            unwrappedRecord.topic(), 
+            unwrappedRecord.kafkaPartition(), 
+            unwrappedRecord.keySchema(), 
+            unwrappedRecord.key(), 
+            updatedSchema, 
+            updatedValue, 
+            unwrappedRecord.timestamp());
+    }
+
+    private Schema makeUpdatedSchema(Schema schema, Schema sourceSchema, String[] addSourceFields) {
+        final SchemaBuilder builder = SchemaUtil.copySchemaBasics(schema, SchemaBuilder.struct());
+        // Get fields from original schema
+        for (org.apache.kafka.connect.data.Field field : schema.fields()) {
+            builder.field(field.name(), field.schema());
+        }
+        // Add the requested source fields if they exist in the source record's schema
+        for(String sourceField: addSourceFields) {
+            if(sourceSchema.field(sourceField) != null) {
```

Comment:
One more question on this one: should we rather raise an exception when referencing an non-existent field instead of silently failing? After all it indicates a misconfiguration likely. How to other SMTs deal with this kind of situation, e.g. the `ExtractField` SMT or others that reference existing fields?

Reply:
I was wondering the same thing, but I saw your comment 

> some fields are optional

and I wasn't sure if that meant they would exist for some records produced by a connector, but not for others. I opted to just not include the field if it didn't exist in the source structure, but if the schema of the source structure will always be consistent for records produced by the same connector I agree that it's better to raise an exception right away.

Reply:
If a field is optional, it will still be present in the _schema_. It may only be that an actual value is present or not present for that field in a given message. Whether the schema is accessible for the SMT or not depends on the configuration and whether the SMT is applied for a source or sink connector. If you are on the sink side and JSON is used without schema, then it wouldn't be present.

So if the schema _is_ available, then you could detect it when a non-existing (in the schema) field is referenced and raise an exception. Without schema you couldn't tell apart a non-existing field from an optional non-present one. Could you check how other SMTs handle that situation? Perhaps raise an exception in the first case but ignore in the second?

Reply:
Ok - I looked at what the InsertField transform does, and updated ExtractNewRecordState to throw a `org.apache.kafka.common.config.ConfigException` if the configuration contains a field that doesn't exist in the source struct's schema.

As far as the case when we don't have the schema: this transform isn't designed to work if the record's value doesn't have a schema (https://github.com/debezium/debezium/blob/1d72a2f8a746b722f09b8dca99bc7b7c78e03a51/debezium-core/src/main/java/io/debezium/transforms/ExtractNewRecordState.java#L114), so we don't need to handle that situation in the `addSourceFields` method as we'll never get there.